### PR TITLE
Add metrics telemetry for dropped logs

### PR DESCRIPTION
**Pull request by gh123man**
### What does this PR do?

Expose user facing metrics when logs are dropped by the agent. 
The agent only drops logs during specific exceptional scenarios. While these are uncommon, we want to provide visibility for them. 

### Describe how to test/QA your changes

In order to test this, we need to simulate intake failures. Since this can be tricky to get right, I put together a docker compose for you to use.

#### How this setup works

the main destination is pointing at a server that only returns `403` forbidden which we do not retry. As a result the destination should drop these logs. 

the additional endpoint points to a server that returns `404` not found. We retry these errors when the endpoint is reliable. The buffers will quickly fill up and the sender will start dropping logs because the main endpoint is unblocked (since those logs are being dropped later on). 

Don't forget to fill in your API key and docker image:

```yaml

version: "3.9"
services:
  agent:
    image: datadog/<agent RC iamge>
    pid: "host"
    environment:
      DD_API_KEY: $API_KEY
      DD_HOSTNAME: qa-logs-dropped-metrics
      DD_LOG_LEVEL: "debug"
      DD_LOGS_ENABLED: "true"
      DD_LOGS_CONFIG_LOGS_DD_URL: "logs-forbidden:8080"
      DD_LOGS_CONFIG_LOGS_NO_SSL: "true"
      DD_LOGS_CONFIG_FORCE_USE_HTTP: "true"
      DD_LOGS_CONFIG_ADDITIONAL_ENDPOINTS: '[{"api_key": "apiKey2", "Host": "logs-notfound", "Port": 8080, "is_reliable": true}]'

    volumes:
      - /var/run/docker.sock:/host/var/run/docker.sock:ro

  logs-emit:
    image: bash
    labels:
      com.datadoghq.ad.logs: '[{"source": "bash-logs", "service": "bash-logs"}]'
    command: 
      - bash
      - -c
      - i=0; while true; do echo "HELLO $i at" $(date); i=$((i+1)); sleep 0.001; done

  logs-forbidden:
    image: python:3.8-slim
    volumes:
      - ./app:/app
    working_dir: /app
    command: python forbidden.py

  logs-notfound:
    image: python:3.8-slim
    volumes:
      - ./app:/app
    working_dir: /app
    command: python notfound.py
```

Also create an `app` folder with the following contents: 

`forbidden.py`
```python3
from http.server import BaseHTTPRequestHandler, HTTPServer

class ErrorHandler(BaseHTTPRequestHandler):
    def do_POST(self):
        self.send_response(403)
        self.send_header('Content-type', 'text/html')
        self.end_headers()
        self.wfile.write(b'403 Forbidden')

if __name__ == '__main__':
    server_address = ('', 8080)
    httpd = HTTPServer(server_address, ErrorHandler)
    httpd.serve_forever()
```

`notfound.py`
```python3
from http.server import BaseHTTPRequestHandler, HTTPServer

class ErrorHandler(BaseHTTPRequestHandler):
    def do_POST(self):
        self.send_response(404)
        self.send_header('Content-type', 'text/html')
        self.end_headers()
        self.wfile.write(b'404 Not Found')

if __name__ == '__main__':
    server_address = ('', 8080)
    httpd = HTTPServer(server_address, ErrorHandler)
    httpd.serve_forever()
```

Once this is up and running, ensure your host is sending metrics.

Query the following metrics and make sure you see them: 

`sum:datadog.agent.logs_client_http_destination.payloads_dropped{*}.as_count()`
`sum:datadog.agent.logs_sender.payloads_dropped{*}.as_count()`
`sum:datadog.agent.logs_sender.messages_dropped{*}.as_count()`

### CODE REVIEW

**Review by pgimalac — APPROVED** [review]

**Review by rayz — APPROVED** [review]

File: pkg/logs/sender/sender.go [hunk 1]
```diff
@@ -92,7 +97,9 @@ func (s *Sender) run() {
 			if !destSender.lastSendSucceeded {
 				if !destSender.NonBlockingSend(payload) {
 					tlmPayloadsDropped.Inc("true", strconv.Itoa(i))
+					telemetry.GetStatsTelemetryProvider().CountNoIndex(payloadsDroppedTelemetryMetricName, 1, []string{"reliable:true"})
 					tlmMessagesDropped.Add(float64(len(payload.Messages)), "true", strconv.Itoa(i))
+					telemetry.GetStatsTelemetryProvider().CountNoIndex(logsDroppedTelemetryMetricName, float64(len(payload.Messages)), []string{"reliable:true"})
```
Comment (jszwedko) [inline pkg/logs/sender/sender.go:102]: Just noticing that we seem to be using two separate patterns for emitting metrics. To satisfy my ignorance, when do you use `telemetry.GetStatsTelemetryProvider().X` vs. initializing a variable like `tlmPayloadsDropped`?

Comment (gh123man) [reply pkg/logs/sender/sender.go:102]: `GetStatsTelemetryProvider` sends metrics directly to the intake - ex: users can query them in their org. 
agent (internal) telemetry is exposed as prometheus metrics and is not collected by default (requires an open metrics check to be setup - very verbose telemetry)

We expose a very limited subset of metrics via "stats telemetry" when they may be directly useful to users.

Comment (jszwedko) [reply pkg/logs/sender/sender.go:102]: Aha, gotcha, that distinction makes sense: whether the metrics are made immediately available to customers that upgrade vs. just published to the internal metrics store and available via the open metrics endpoint. In this case we think these metrics are high value enough to publish immediately? I'm assuming based on the analogous dropped points metric we have for metrics publishing?

Comment (gh123man) [reply pkg/logs/sender/sender.go:102]: > In this case we think these metrics are high value enough to publish immediately? I'm assuming based on the analogous dropped points metric we have for metrics publishing?
Yep exactly

**Review by jszwedko — APPROVED** [review]
Thanks for the detailed testing instructions! I was able to run this and observe the new metrics.

File: pkg/logs/sender/sender.go (hunk 1 above)
Comment (vickenty) [reply pkg/logs/sender/sender.go:102]: We do have a facility to send prometheus metrics by default (e.g. used [here](https://github.com/datadog/datadog-agent/blob/cc21ca40c5c2f69312b5899ba5e6ff3bad912264/comp/forwarder/defaultforwarder/internal/retry/point_count_telemetry.go#L16)). Would it make sense to use it here?

Comment (gh123man) [reply pkg/logs/sender/sender.go]: Thanks! Somehow I missed the PR that introduced this. Much better solution 👍

**Review by vickenty — APPROVED** [review]

**Pull request merged**